File: apps/charter-app/src/domain/pairingSas.test.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";
import { nip19 } from "nostr-tools";
import { pairingSas } from "./pairingSas";

const PK = "a".repeat(63) + "b";
const PK2 = "c".repeat(64);
const TOKEN = "0123456789abcdef0123456789abcdef";

/*
 * The frozen cross-language vectors (S6). THREE runtimes derive this code —
 * this module, the hand-written copy in `public/pair/index.html`, and the ward
 * app's Kotlin `PairingSas` — because they cannot share code. The failure to
 * guard against is not "the code is wrong", it is "the screens disagree",
 * which teaches a family the check is broken and to tap through it. All three
 * are pinned to this one file; the Kotlin side reads it off the shared test
 * classpath (`android/app/build.gradle.kts` sourceSets).
 */
const VECTORS = JSON.parse(
  readFileSync(
    resolve(__dirname, "../../../../core/crates/charter-testkit/vectors/pairing/sas_vectors.json"),
    "utf8",
  ),
) as {
  vectors: { name: string; pubkeyHex: string; token: string; code: string }[];
  npubVectors: { pubkeyHex: string; npub: string }[];
};

describe("the frozen cross-language vectors", () => {
  it("has the full set — a truncated file must fail, not assert nothing", () => {
    expect(VECTORS.vectors.length).toBeGreaterThanOrEqual(6);
    expect(VECTORS.npubVectors.length).toBeGreaterThanOrEqual(4);
  });

  it("pairingSas agrees with every one", async () => {
    for (const v of VECTORS.vectors) {
      expect(await pairingSas(v.pubkeyHex, v.token), v.name).toBe(v.code);
    }
  });

  it("nip19.npubEncode agrees with every one", () => {
    for (const v of VECTORS.npubVectors) {
      expect(nip19.npubEncode(v.pubkeyHex)).toBe(v.npub);
    }
  });
});

describe("pairingSas", () => {
  it("is six digits in two groups", async () => {
    expect(await pairingSas(PK, TOKEN)).toMatch(/^\d{3} \d{3}$/);
  });

  it("is stable for the same guardian and token", async () => {
    expect(await pairingSas(PK, TOKEN)).toBe(await pairingSas(PK, TOKEN));
  });

  it("does not care how the pubkey was cased", async () => {
    expect(await pairingSas(PK.toUpperCase(), TOKEN)).toBe(await pairingSas(PK, TOKEN));
  });

  it("changes when the guardian changes — the attack it exists for", async () => {
    expect(await pairingSas(PK2, TOKEN)).not.toBe(await pairingSas(PK, TOKEN));
  });

  // A code tied only to the pubkey would repeat for every pairing this
  // guardian ever does, so one glimpse of it would be replayable forever.
  it("changes when the token changes, so a code seen once cannot be reused", async () => {
    expect(await pairingSas(PK, "f".repeat(32))).not.toBe(await pairingSas(PK, TOKEN));
  });
});

/*
 * The anti-drift check (S6). `public/pair/index.html` is static HTML with no
 * bundler, so it carries its own copy of the derivation. Two screens showing
 * DIFFERENT codes for the same link would be worse than showing none — it
 * would teach a family that the check is broken and to tap through it. So the
 * page's own function is pulled out of the file and run head-to-head here.
 */
describe("the /pair page's copy of the derivation", () => {
  const html = readFileSync(
    resolve(__dirname, "../../public/pair/index.html"),
    "utf8",
  );

  /** Lift `function sasFrom(…) { … }` out of the page and make it callable. */
  function pageSasFrom(): (pk: string, token: string) => Promise<string> {
    const start = html.indexOf("function sasFrom(");
    expect(start, "the /pair page must still define sasFrom").toBeGreaterThan(-1);
    // Balance braces from the function's opening `{` to its close.
    let depth = 0;
    let end = -1;
    for (let i = html.indexOf("{", start); i < html.length; i++) {
      if (html[i] === "{") depth++;
      else if (html[i] === "}" && --depth === 0) {
        end = i + 1;
        break;
      }
    }
    expect(end, "sasFrom must be a balanced function body").toBeGreaterThan(-1);
    const src = html.slice(start, end);
    return new Function(`${src}; return sasFrom;`)() as (
      pk: string,
      token: string,
    ) => Promise<string>;
  }

  it("agrees with src/domain/pairingSas.ts", async () => {
    const fromPage = pageSasFrom();
    for (const [pk, token] of [
      [PK, TOKEN],
      [PK2, TOKEN],
      [PK, "f".repeat(32)],
      [PK, ""],
    ] as const) {
      expect(await fromPage(pk, token)).toBe(await pairingSas(pk, token));
    }
  });

  it("agrees with the frozen vectors the Kotlin side is pinned to", async () => {
    const fromPage = pageSasFrom();
    for (const v of VECTORS.vectors) {
      expect(await fromPage(v.pubkeyHex, v.token), v.name).toBe(v.code);
    }
  });
});

/*
 * And the npub the page renders must be the SAME string Kintrinsic shows the
 * parent — the page hand-rolls bech32 because it cannot import nostr-tools.
 */
describe("the /pair page's npub encoder", () => {
  const html = readFileSync(
    resolve(__dirname, "../../public/pair/index.html"),
    "utf8",
  );

  function pageNpubEncode(): (hex: string) => string {
    const start = html.indexOf("function npubEncode(");
    expect(start, "the /pair page must still define npubEncode").toBeGreaterThan(-1);
    let depth = 0;
    let end = -1;
    for (let i = html.indexOf("{", start); i < html.length; i++) {
      if (html[i] === "{") depth++;
      else if (html[i] === "}" && --depth === 0) {
        end = i + 1;
        break;
      }
    }
    const src = html.slice(start, end);
    return new Function(`${src}; return npubEncode;`)() as (hex: string) => string;
  }

  it("matches nip19.npubEncode", () => {
    const encode = pageNpubEncode();
    for (const pk of [PK, PK2, "0".repeat(64), "f".repeat(64), "0123456789abcdef".repeat(4)]) {
      expect(encode(pk)).toBe(nip19.npubEncode(pk));
    }
  });

  it("matches the frozen npub vectors the Kotlin side is pinned to", () => {
    const encode = pageNpubEncode();
    for (const v of VECTORS.npubVectors) {
      expect(encode(v.pubkeyHex)).toBe(v.npub);
    }
  });
});

File: android/app/src/test/kotlin/org/forgesworn/charter/ui/PairingSasTest.kt
package org.forgesworn.charter.ui

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The ward app's pairing code, pinned to the SAME frozen vectors the guardian
 * app and the `/pair` page are pinned to (S6, review 2026-08-07).
 *
 * Three runtimes implement this derivation because they cannot share code. The
 * failure this guards against is not "the code is wrong" — it is "the two
 * screens disagree", which teaches a family the check is broken and to tap
 * through it. A one-sided edit to any implementation fails here.
 */
class PairingSasTest {

    private fun vectors(): JSONObject {
        val stream = object {}.javaClass.classLoader!!
            .getResourceAsStream("pairing/sas_vectors.json")
            ?: error("shared pairing SAS vectors not found on the test classpath")
        return JSONObject(stream.bufferedReader().use { it.readText() })
    }

    @Test
    fun `the code matches the shared cross-language vectors`() {
        val arr = vectors().getJSONArray("vectors")
        // Completeness, not just non-emptiness: a truncated or unreadable
        // vectors file must fail rather than quietly assert nothing.
        assertTrue("expected the full vector set", arr.length() >= 6)
        for (i in 0 until arr.length()) {
            val v = arr.getJSONObject(i)
            assertEquals(
                v.getString("name"),
                v.getString("code"),
                PairingSas.code(v.getString("pubkeyHex"), v.getString("token")),
            )
        }
    }

    @Test
    fun `the npub matches the shared cross-language vectors`() {
        val arr = vectors().getJSONArray("npubVectors")
        assertTrue("expected the full npub vector set", arr.length() >= 4)
        for (i in 0 until arr.length()) {
            val v = arr.getJSONObject(i)
            assertEquals(v.getString("npub"), PairingSas.npub(v.getString("pubkeyHex")))
        }
    }

    @Test
    fun `always six digits in two groups`() {
        for (i in 0 until 200) {
            val code = PairingSas.code("%064x".format(i), "t$i")
            assertTrue(code, Regex("^\\d{3} \\d{3}$").matches(code))
        }
    }

    /* A code tied only to the pubkey would repeat for every pairing this
     * guardian ever does, so one glimpse of it would be replayable forever. */
    @Test
    fun `a fresh token gives a fresh code`() {
        val pk = "a".repeat(64)
        assertTrue(PairingSas.code(pk, "one") != PairingSas.code(pk, "two"))
    }

    /* THE case this exists for: a stranger's link must not produce the code
     * the grown-up's screen is showing. */
    @Test
    fun `a different guardian gives a different code`() {
        val token = "0123456789abcdef0123456789abcdef"
        assertTrue(PairingSas.code("a".repeat(64), token) != PairingSas.code("b".repeat(64), token))
    }

    // ---- what the ward's confirm screen actually shows -------------------

    private val LINK =
        "bunker://aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab" +
            "?relay=wss://relay.example&kind=charter&token=0123456789abcdef0123456789abcdef"

    @Test
    fun `reads the code and the guardian id out of a real pairing link`() {
        val check = PairingSas.check(LINK)!!
        assertEquals("895 208", check.code)
        assertEquals(
            "npub1424242424242424242424242424242424242424242424242424sg6tqgp",
            check.npub,
        )
    }

    @Test
    fun `the App Link fragment form carries the same identity as the raw one`() {
        // MainActivity unwraps `https://…/pair#bunker://…` to the raw form
        // before this sees it; both must land on the same check.
        assertEquals(PairingSas.check(LINK), PairingSas.check(LINK))
    }

    @Test
    fun `a link with no token still shows a code`() {
        // The cable/paste path can arrive tokenless. There is still a guardian
        // identity to check, and it is still worth checking.
        val bare = LINK.substringBefore("&token=")
        assertEquals("278 600", PairingSas.check(bare)!!.code)
    }

    /*
     * "Say nothing rather than something reassuring." A half-rendered check —
     * or one left over from a previous link — is a check that passes for the
     * wrong thing.
     */
    @Test
    fun `an unreadable link produces no check at all`() {
        for (bad in listOf(
            "",
            "   ",
            "not a link",
            "https://charter.mysignet.app/pair",
            "bunker://",
            "bunker://short?token=x",
            "bunker://" + "z".repeat(64),
            "javascript:alert(1)",
        )) {
            assertNull(bad, PairingSas.check(bad))
        }
    }

    /* THE reason this is driven off the text field and not off the scan: a
     * ward who edits the link must not be left looking at the old code. */
    @Test
    fun `a different link gives a different check`() {
        val other = LINK.replace("aaaa", "bbbb")
        assertTrue(PairingSas.check(LINK)!!.code != PairingSas.check(other)!!.code)
        assertTrue(PairingSas.check(LINK)!!.npub != PairingSas.check(other)!!.npub)
    }

    @Test
    fun `a malformed pubkey yields no npub rather than a plausible one`() {
        assertNull(PairingSas.npub(""))
        assertNull(PairingSas.npub("a".repeat(63)))
        assertNull(PairingSas.npub("a".repeat(65)))
        assertNull(PairingSas.npub("z".repeat(64)))
    }
}
